### PR TITLE
Application insights config

### PR DESCRIPTION
**Pull request by doncem**
### JIRA link (if applicable) ###

[Spike: Centralised Logging in Azure](https://tools.hmcts.net/jira/browse/RPE-232)

### Change description ###

Add configuration file for app insights. Request filter requires agent to be included into app run. Unfortunately it is not in maven and have to manually download it ¯\_(ツ)_/¯. Microsoft

Here's a successful log from spring boot in docker:
```bash
AI: INFO 25-01-2018 15:55, 14: Registering WebApp with name 'Send Letter Provider'
AI: INFO 25-01-2018 15:55, 14: Configuration file has been successfully found as resource
AI: INFO 25-01-2018 15:55, 14: 'MaxTelemetryBufferCapacity': null value is replaced with '500'
AI: INFO 25-01-2018 15:55, 14: 'FlushIntervalInSeconds': null value is replaced with '5'
AI: TRACE 25-01-2018 15:55, 14: Using Http Client version 4.3+
AI: INFO 25-01-2018 15:55, 14: 'Channel.MaxTransmissionStorageCapacityInMB': null value is replaced with '10'
AI: INFO 25-01-2018 15:55, 16: Starting to poll for Docker context file under: /usr/appinsights/docker/docker.info
AI: INFO 25-01-2018 15:55, 14: Configuration file has been successfully found as resource
AI: INFO 25-01-2018 15:55, 14: 'MaxTelemetryBufferCapacity': null value is replaced with '500'
AI: INFO 25-01-2018 15:55, 14: 'FlushIntervalInSeconds': null value is replaced with '5'
AI: INFO 25-01-2018 15:55, 14: 'Channel.MaxTransmissionStorageCapacityInMB': null value is replaced with '10'
AI: INFO 25-01-2018 15:55, 18: Starting to poll for Docker context file under: /usr/appinsights/docker/docker.info
AI: TRACE 25-01-2018 15:55, 14: InstrumentationKeyResolver - pending resolution of instrumentation key: 00000000-0000-0000-0000-000000000000
AI: TRACE 25-01-2018 15:55, 14: InstrumentationKeyResolver - pending resolution of instrumentation key: 00000000-0000-0000-0000-000000000000
AI: INFO 25-01-2018 15:55, 14: Registered WebApp 'Send Letter Provider' key='Send Letter Provider'
AI: INFO 25-01-2018 15:55, 14: Successfully registered the filter 'ApplicationInsightsWebFilter'
```

P.S. you can see duplicate logs because request filter creates it's own telemetry client

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -107,12 +111,13 @@ dependencies {
...
   compile group: 'uk.gov.hmcts.reform', name: 'http-proxy-spring-boot-autoconfigure', version: '1.1.0'
-  compile group: 'uk.gov.hmcts.reform', name: 'service-auth-provider-client', version: '0.4.1'
+  compile group: 'uk.gov.hmcts.reform', name: 'service-auth-provider-client', version: '0.4.2'
 
   compile group: 'org.springframework.cloud', name: 'spring-cloud-starter-hystrix', version: '1.4.0.RELEASE'
   compile group: 'org.springframework.cloud', name: 'spring-cloud-starter-hystrix-dashboard', version: '1.4.0.RELEASE'
 
-  compile group: 'com.microsoft.azure', name: 'applicationinsights-web', version: '1.0.10'
+  compile group: 'com.microsoft.azure', name: 'applicationinsights-agent', version: '2.0.0-BETA'
+  compile group: 'com.microsoft.azure', name: 'applicationinsights-web', version: '2.0.0-BETA'
```
Comment (rkondratowicz) [inline build.gradle]: bleeding edge ;)

File: src/main/resources/ApplicationInsights.xml [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+    <TelemetryModules>
+        <Add type="com.microsoft.applicationinsights.web.extensibility.modules.WebRequestTrackingTelemetryModule"/>
+        <Add type="com.microsoft.applicationinsights.web.extensibility.modules.WebSessionTrackingTelemetryModule"/>
+        <Add type="com.microsoft.applicationinsights.web.extensibility.modules.WebUserTrackingTelemetryModule"/>
+    </TelemetryModules>
+
+    <TelemetryInitializers>
+        <!-- core -->
+        <Add type="com.microsoft.applicationinsights.extensibility.initializer.docker.DockerContextInitializer"/>
```
Comment (rkondratowicz) [inline src/main/resources/ApplicationInsights.xml]: should this be in main package?

File: build.gradle (hunk 1 above)
Comment (doncem) [reply build.gradle]: milestone seemed pretty close: 97%. just dropping java 6 support :D. the config just doesn't work for 1.x series lol

**Comment by rkondratowicz** [comment]
Why was the jar added?

**Comment by doncem** [comment]
the request filter could not be initialised due to agent lib not found - need to include it manually since there is no repo available. according to microsoft docs. among those links i found what's missing

File: src/main/java/uk/gov/hmcts/reform/sendletter/config/InsightConfiguration.java [hunk 3]
```diff
@@ -1,14 +1,38 @@
...
 public class InsightConfiguration {
 
     @Bean
     public TelemetryClient getTelemetryClient() {
         return new TelemetryClient();
     }
+
+    @Bean
+    @ConditionalOnWebApplication
+    public FilterRegistrationBean registerWeRequestTrackingFilter(
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/sendletter/config/InsightConfiguration.java]: typo in 'web'

**Comment by rkondratowicz** [comment]
should the instrumentation key be logged?

**Comment by doncem** [comment]
That's my worry too. Could not find how to override app insights logger

**Comment by doncem** [comment]
The log becomes
```bash
AI: INFO 26-01-2018 09:20, 15: Registering WebApp with name 'Send Letter Provider'
AI: INFO 26-01-2018 09:20, 15: Configuration file has been successfully found as resource
AI: INFO 26-01-2018 09:20, 15: Configuration file has been successfully found as resource
AI: INFO 26-01-2018 09:20, 15: Registered WebApp 'Send Letter Provider' key='Send Letter Provider'
AI: INFO 26-01-2018 09:20, 15: Successfully registered the filter 'ApplicationInsightsWebFilter'
```

**Pull request merged**